### PR TITLE
Feature enumerable

### PR DESCRIPTION
**Pull request by Dipeshtwis**
This project follows

- the git-flow method
- made all the method of our own

### CODE REVIEW

**Comment by Dipeshtwis** [comment]
I enabled the stickler but still, the stickler is not checking the errors. Maybe because I have changed the name of the repository. I already ran on my local machine and there was no offence detected.

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
+module Enumerable
+  def my_each
+    i = 0
+    while i < length
+      yield(self[i])
+      i += 1
+
+    end
+    self
+  end
```
Comment (MarioDena) [inline main.rb]: - [x] #my_each should return an Enumerator if no block is given
- [x] when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] my_each should, when a block is given when self is a range behaves like #each calls the given block once for each element in self
- [x] my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self

This mostly means you have to deal with ranges and hashes correctly

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_each_with_index
+    i = 0
+    while i < length
+      yield(self[i], i)
+      i += 1
+
+    end
+    self
+  end
```
Comment (MarioDena) [inline main.rb]: Same problem that we have on my_each.

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_select
+    selected_item = []
+    my_each do |i|
+      selected_item.push(i)
+    end
+    selected_item
+  end
```
Comment (MarioDena) [inline main.rb]: - [x] #my_select should return an array containing all elements of enum for which the given block returns a true value
- [x] #my_select should return an array containing all elements of enum for which the given block returns a true value::range
- [x] #my_select should return an enumerator if no block is given

This is a problem with the way you are returning data.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_all?
+    my_each do |i|
+      return false if yield(i) == false
+    end
+    true
+  end
+
```
Comment (MarioDena) [inline main.rb]: - [x] #my_all should return false if the block returns false or nil::range
- [x] #my_all? when no block or argument is given returns true when none of the collection members are false or nil
- [x] #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [x] #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
- [x] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

What happens here is that you need to deal with regex and classes and also the way you return data.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  def my_any?
+    my_each do |i|
+      return true if yield(i) == true
+    end
+    false
+  end
```
Comment (MarioDena) [inline main.rb]: - [x] #my_any? returns false if the block ever returns a value false or nil::range
- [x] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] #my_any? when no block or argument is given returns false if at least one of the collection is not true

Remember all the problems with regex and classes apply to this method too.

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_none?
+    my_each do |i|
+      return false if yield(i) == true
+    end
+    true
+  end
+
```
Comment (MarioDena) [inline main.rb]: Remember pretty much all errors present on #my_any? are present here on my_none?

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  def my_count?
+    count = 0
+    my_each do |i|
+      count += 1 if yield(i) == true
+    end
+    count
+  end
```
Comment (MarioDena) [inline main.rb]: - [x]  #my_count should return the number of items in enum through enumeration
- [x]  #my_count should return the number of items in enum through enumeration::range
- [x]  #my_count counts the number of items in enum that are equal to item if an argument is given
- [x]  #my_count counts the number of elements yielding a true value if a block is given
- [x]  #my_count does not mutate the original array

The most important thing here is that you must not mutate the original array, you must return a result depending on the input variables but the original array should not mutate.

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_map
+    items = []
+    my_each do |i|
+      items << (proc.nil? ? proc.call(i) : yield(i))
+    end
+    items
+  end
```
Comment (MarioDena) [inline main.rb]: - [x] #my_map should return a new array with the results of running a given block once for every element in enum::range
- [x]  #my_map should return an Enumerator if no block is given
- [x]  #my_map should execute only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

Now here you should be very careful when both a block and a proc are given, Ruby will trow an error but I guarantee you it is possible. Just search for example on stack-overflow

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_inject(first = self[0])
+    my_each do |i|
+      first = yield(first, i)
+    end
+    first
+  end
+end
```
Comment (MarioDena) [inline main.rb]: - [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block
- [x] #my_inject when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
- [x] #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
- [x] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
- [x] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range

Now here you just have to watch out for how your method works with blocks symbols, and ranges.

**Review by MarioDena — CHANGES_REQUESTED** [review]
Hello! Good work on your project, however most methodsare failing to behaive like their ruby counterparts.

## Changes required 🔴

**Comment by ignatius22** [comment]
Updated project includes : 
Enumerable was returned in absence of block.
conditions for regex and class were included.
Original array was made to be immutable.

**Review by rOluochKe — APPROVED** [review]
Hello Team!
### Good job you have done and for making previous review changes  :clap: The project has been approved. Congratulations :boom: :100:
### You can merge and submit a project completion form from the dashboard. Good luck with the next project.
- Happy coding!

**Pull request merged**